### PR TITLE
Autonomous filter category search

### PR DESCRIPTION
**Pull request by kiruys**
Features of the filtertree (in agreement with Jelmer): 

- The filtertree is fed by the search endpoint instead of the filter-categories endpoint.

- It functions as a dumb component (i.e., doesn't depend on the vuex store)

- Filteritems are shown when count > 0

- Filteritems are sorted based on count.

- A top 3 of Filteritems is shown.

- If there are more than 3 Filteritems with a count > 0, there is a button that can toggle more items.

- Selected filters are shown, and the category is opened.

- Selected filters can be deselected.

- When a date range has been selected, the date range component is opened.

- And most importantly: Filter counts are shown.

NB 1: Because the search-endpoint is used and filter-items are only shown when count > 0, 'gebruiksrechten', 'bruikbaar als', and 'taal' are currently not displayed.

NB2: If we want to show counts of filter-items of a filter-category that already has a selected filter-item (ala Bol.com) that should be added in the search endpoint.

### CODE REVIEW

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 1]
```diff
@@ -7,24 +6,26 @@ export default {
...
+  props: ['filterCategories', 'materials', 'selectedFilters'],
   methods: {
     generateSearchMaterialsQuery,
     hasVisibleChildren(category) {
       if (!category.children.length) {
         return false
       }
       return _.some(category.children, child => {
-        return !child.is_hidden
+        if (child.count > 0) {
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:26]: `return child.count > 0 && !child.is_hidden`

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 2]
```diff
@@ -36,76 +37,47 @@ export default {
...
-        }
-
-        filters = Object.assign({}, this.value, {
-          filters
-        })
-
-        this.$router.push(this.generateSearchMaterialsQuery(filters))
+    onChange(e) {
+      const { categoryId, itemId } = e.target.dataset
+      const filter = { external_id: categoryId, items: [itemId] }
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:42]: Kan je niet meerdere items in een filter hebben?

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 3]
```diff
@@ -36,76 +37,47 @@ export default {
...
+      let filters = this.selectedFilters
 
-        this.$emit('input', filters)
-
-        return filters
+      if (e.target.checked) {
+        filters.push(filter)
+      } else {
+        filters.splice(filters.indexOf(filter), 1)
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:48]: Werkt dit op objecten? En ik vind een filter duidelijker, waar je hem eruit filtert (maar dat kan ook persoonlijke voorkeur zijn).

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 4]
```diff
@@ -36,76 +37,47 @@ export default {
...
       }
+      this.executeSearch(filters)
     },
-    onChange(event, parent) {
-      if (!_.isNil(parent)) {
-        parent.selected = !parent.selected
+    onDateChange(dates) {
+      const { start_date, end_date } = dates
+      const filter = {
+        external_id: this.publisherDate,
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:55]: die variabele moet dus eigenlijk `publisherDateExternalId` heten?

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 5]
```diff
@@ -36,76 +37,47 @@ export default {
...
+        items: [start_date, end_date]
       }
-      this.$store.commit('SET_FILTER_SELECTED', event.target.dataset.categoryId)
-      this.executeSearch()
-    },
-    onDateChange() {
-      this.executeSearch()
+      let filters = this.selectedFilters.filter(
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:58]: kan ook const

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 6]
```diff
@@ -36,76 +37,47 @@ export default {
...
+      this.executeSearch(filters)
     },
-    executeSearch() {
-      let searchText = this.$store.getters.materials.search_text
-      let ordering = this.$store.getters.materials.ordering
+    executeSearch(filters = []) {
+      const { ordering, search_text } = this.materials
       let searchRequest = {
-        search_text: searchText,
+        search_text: search_text,
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:67]: hm, niet searchText is meer javascript naming. En anders zijn die dubbele punten niet nodig, want de object keys zijn hetzelfde als de variabele naam

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 7]
```diff
@@ -128,21 +95,45 @@ export default {
...
-     * @returns {*}
-     */
     filtered_categories() {
-      // Return categories which builds the filter tree
-      return this.filter_categories
-        ? _.filter(this.$store.getters.filter_categories.results, {
-            is_hidden: false
-          })
+      // Return categories that build the filter tree
+      let filteredCategories = this.filterCategories
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:116]: kunnen we die this.filterCategories niet default `[]` laten zijn? Neemt een hoop condities weg.

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 8]
```diff
@@ -128,21 +95,45 @@ export default {
...
+        ? this.filterCategories.filter(item => item.is_hidden === false)
         : []
+
+      // set selected filters
+      if (this.selectedFilters) {
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:121]: same here

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 9]
```diff
@@ -128,21 +95,45 @@ export default {
...
+        let selectedItems = []
+        this.selectedFilters.forEach(filter => {
+          selectedItems = selectedItems.concat(filter.items)
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:124]: hier doe je een `flatMap`

File: service/surf/apps/materials/views.py [hunk 10]
```diff
@@ -150,9 +151,26 @@ def post(self, request, *args, **kwargs):
                 "items": self.parse_theme_drilldowns(discipline_items)
             })
 
+        drill_down_dict = {item['external_id']: item for item in res["drilldowns"]}
```
Comment (Ruben-Hartog) [inline service/surf/apps/materials/views.py:154]: 🤯

File: portal/src/components/DatesRange/DatesRange.component.js [hunk 11]
```diff
@@ -12,29 +19,23 @@ export default {
     this.popupItem = this.$el
   },
   data() {
-    const { value } = this
+    const { datesFilter } = this
```
Comment (Ruben-Hartog) [inline portal/src/components/DatesRange/DatesRange.component.js:22]: Waarom heb je dit veranderd? Kunnen we niet de value blijven gebruiken? Hoe werkt die `$emit` nu? Die werkt toch met v-model?

File: portal/src/components/Collections/AddMaterialPopup.vue [hunk 12]
```diff
@@ -41,7 +41,7 @@
 import { isEmpty, map } from 'lodash'
 import { mapGetters } from 'vuex'
 import Popup from '~/components/Popup'
-import Search from '~/components/FilterCategories/Search'
+import Search from '~/components/Search'
```
Comment (Ruben-Hartog) [inline portal/src/components/Collections/AddMaterialPopup.vue:44]: 👍

**Comment by fako** [comment]
Voordat ik de code ga reviewen even wat dingen die me opvallen als ik handmatig test

**Comment by fako** [comment]
Steps to reproduce:
- Select a filter

Expected:
- Category of the filter remains open
- Selection of the filter remains selected

Actual:
- Category closes and I need to re-open. Not sure if people (Jelmer) will except this. It's rather onorthodox I think.
- Filter is not selected. This makes it impossible to deselect a filter. This is a serious bug. You can't expect people to work around this

Proposed solution:
- Read the query parameters in the backend and push them into serializer context. Then create is_selected/selected serializer field and return if things are selected. This makes it testable in Python. We can remove the reading of query parameters from the frontend. Writing the parameters remains necessary
- Keep track of open categories in an object and every time we get new filter categories update the "is_open" state inside the component (use Vue watchers). Note that component should open any categories that have is_selected/selected filters upon mount (will write separate bug to explain). This is hard to do within Python without updating the query parameter protocol, so I wouldn't go there.

Alternative solution:
- Keep track of selection with an object in the component and update selection whenever new filter categories arrive (use Vue watchers). Basically same as is_open state. This would make the frontend component less dumb and less testable.

**Comment by fako** [comment]
Steps to reproduce:
- Select text as file type. Search for "veel"

Expected:
- "soort materiaal" category is open. 
- It shows all filters and "tekst" is selected

Actual:
- "soort materiaal" is closed
- All filters are gone except "tekst". It is not selected

Proposed solution:
- If the "is_selected" property comes from the backend. Then upon mount the component needs to scan the incoming filter categories. If any child of the category is selected then it should set is_open/isOpen. Whenever a new search gets done on the same page and new filter categories enter the component then it should not re-open manually closed categories. So only set is_open/isOpen on mount and leave it for the rest to the user (except when he leaves pages and comes back). On production it is even the case that open categories remain open after visiting a search result and then coming back (which triggers a mount). We can keep this behaviour if you store is_open/isOpen in a store instead of the component. Clear the store when making a new search.

Alternative solution:
- When dealing with is_selected/selected from the frontend side it will act the same as above except that mount should also check for selected state not just open state.

**Comment by fako** [comment]
Steps to reproduce:
- Search for "veel"
- Open "gebruiksrechten"

Expected:
- Gebruiksrechten should show filters

Actual:
- There a no filters in gebruiksrechten

Important:
- Unable to test this, but when I select a gebruiksrecht filter I should see that it filters on at least 3 items. Each gebruiksrecht has 2 hidden children and those should get selected whenever I select a parent filter. This is necessary to find *all* items that have that copyright, including version 3 and 4 of the same license.

**Comment by fako** [comment]
Steps to reproduce:
- Search for "veel"
- Open "publicatiedatum"
- Select a date

Expected:
- Materials are filtered by publication date (either "from", "until" or both)

Actual:
- Frontend crashes

File: portal/src/components/DatesRange/DatesRange.component.js (hunk 11 above)
Comment (fako) [reply portal/src/components/DatesRange/DatesRange.component.js:22]: Ja, lijkt inderdaad kapot. Zie comments

File: portal/src/components/FilterCategories/FilterCategories.component.html [hunk 13]
```diff
@@ -17,63 +17,40 @@ <h3 class="filter-categories__title">{{ $t('Filter') }}</h3>
...
+          >
+            <div class="filter-checkbox">
               <input
                   type="checkbox"
                   :id="item.external_id"
                   :value="item.external_id"
-                  :data-category-id="item.external_id"
+                  :data-category-id="category.external_id"
+                  :data-item-id="item.external_id"
                   v-model="filtered_categories[index].children[indexItem].selected"
```
[the remark this answers is not in the document]
Comment (fako) [reply portal/src/components/FilterCategories/FilterCategories.component.html:46]: Misschien wel. Lukte me eerder niet, maar aangezien component nu dom gaat worden kan het misschien gaan werken.

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 14]
```diff
@@ -7,24 +6,26 @@ export default {
...
     }
   },
+  props: ['filterCategories'],
   methods: {
     generateSearchMaterialsQuery,
     hasVisibleChildren(category) {
       if (!category.children.length) {
         return false
       }
       return _.some(category.children, child => {
```
[the remark this answers is not in the document]
Comment (fako) [reply portal/src/components/FilterCategories/FilterCategories.component.js:25]: Ja, kan inderdaad wel. Lodash some is cross-browser, maar daar geven we minder om nu we een scherm hebben dat oude browsers blockt 👍

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 1 above)
Comment (fako) [reply portal/src/components/FilterCategories/FilterCategories.component.js:26]: Dit is problematisch. Een child met een count van 0 is niet altijd hidden. Misschien disabled?

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 2 above)
Comment (fako) [reply portal/src/components/FilterCategories/FilterCategories.component.js:42]: Ja, dat kan met bijvoorbeeld "vakgebieden". Deze filter staat momenteel uit, maar lijkt me een goede. Bovendien is het noodzaak voor CC filters (zie comments)

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 3 above)
Comment (fako) [reply portal/src/components/FilterCategories/FilterCategories.component.js:48]: Lodash without/remove to the resque

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 6 above)
Comment (fako) [reply portal/src/components/FilterCategories/FilterCategories.component.js:67]: Django backend geeft default snake case terug, dus zou daar wel een uitzondering voor willen maken, anders moet heel de backend om en ik heb het eerlijk gezegd ook nooit hoeven doen, dus weet niet hoe makkelijk dit gaat. Voor JS variabelen is camel case prima wat mij betreft.

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 7 above)
Comment (fako) [reply portal/src/components/FilterCategories/FilterCategories.component.js:116]: Lijkt me een goed idee.

File: service/surf/apps/materials/views.py (hunk 10 above)
Comment (fako) [reply service/surf/apps/materials/views.py:154]: Het heet dictionary comprehension. Neefje van list comprehension. Standaard Python.

Comment (Ruben-Hartog) [reply service/surf/apps/materials/views.py:154]: Ik bedoelde meer het geheel 😛

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 6 above)
Comment (Ruben-Hartog) [reply portal/src/components/FilterCategories/FilterCategories.component.js:67]: Ja, dat bedoel ik 😉

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (Ruben-Hartog) [reply portal/src/components/FilterCategories/FilterCategories.component.js:25]: Wordt de babel polyfill niet gebruikt? Want die zorgt ervoor dat dit soort dingen cross-browser kunnen. Dan heb je lodash (gelukkig) eigenlijk nooit meer nodig

File: portal/src/components/FilterCategories/FilterCategories.component.html (hunk 13 above)
[the remark this answers is not in the document]
Comment (kiruys) [reply portal/src/components/FilterCategories/FilterCategories.component.html:46]: gek genoeg werkt item.selected niet

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 2 above)
Comment (kiruys) [reply portal/src/components/FilterCategories/FilterCategories.component.js:42]: ja maar als je een item aanzet, dan komt er meteen een search call, dus dan is het er maar 1

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 1 above)
Comment (kiruys) [reply portal/src/components/FilterCategories/FilterCategories.component.js:26]: wil je dan alle children laten zien? ook als ze geen resultaten hebben? wat ik al eerder zei: het is me niet zo duidelijk hoe het moet werken.

File: portal/src/components/DatesRange/DatesRange.component.js (hunk 11 above)
Comment (kiruys) [reply portal/src/components/DatesRange/DatesRange.component.js:22]: het is zeker niet kapot, het werkt nu met props.

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 15]
```diff
@@ -115,34 +93,88 @@ export default {
...
+        item => item.external_id === this.publisherDateExternalId
+      )
+
+      if (datesFilter && datesFilter.items) {
+        const items = datesFilter.items.filter(item => item !== null)
+        return items.length > 0
+      } else return false
+    },
+    sortFilterItems(items) {
+      const nullCounts = items.filter(item => item.count === null)
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:136]: Ik vind nog steeds dat null een `0` moet zijn hier

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 16]
```diff
@@ -115,34 +93,88 @@ export default {
...
+        }
+      } else {
+        return {
+          start_date: null,
+          end_date: null
+        }
+      }
+    },
+    hasSelectedChildren(cat) {
+      return cat.children.filter(child => child.selected === true).length > 0
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:121]: `=== true` is hier overbodig

Comment (Ruben-Hartog) [reply portal/src/components/FilterCategories/FilterCategories.component.js:121]: En ik zou gebruiken maken van: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 17]
```diff
@@ -115,34 +93,88 @@ export default {
...
-        !_.isNil(category.title_translations) &&
-        !_.isEmpty(category.title_translations)
-      ) {
+      if (category && category.title_translations) {
         return category.title_translations[language]
       }
       return category.name
+    },
+    datesRangeFilter() {
+      if (this.selectedFilters) {
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:103]: Kan dit null zijn? Kunnen we het default een leeg array maken? Scheelt weer condities.

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 18]
```diff
@@ -40,72 +40,50 @@ export default {
...
+      const filter = {
+        external_id: this.publisherDateExternalId,
+        items: [start_date, end_date]
       }
-      this.$store.commit('SET_FILTER_SELECTED', event.target.dataset.categoryId)
-      this.executeSearch()
-    },
-    onDateChange() {
-      this.executeSearch()
+      const filters = this.selectedFilters.filter(
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:65]: Ik zou een map doen, waar we altijd het origineel returnen, tenzij het die publisherdate is.

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 19]
```diff
@@ -40,72 +40,50 @@ export default {
...
-          filters
-        })
-
-        this.$router.push(this.generateSearchMaterialsQuery(filters))
-
-        this.$emit('input', filters)
+    onChange(e) {
+      const { categoryId, itemId } = e.target.dataset
+      const filter = { external_id: categoryId, items: [] }
+      const categoryFilter =
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:46]: Hier zou ik ook een map doen. Altijd origineel, tenzij het external id overeen komt

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 20]
```diff
@@ -1,35 +1,35 @@
...
       }
-      return _.some(category.children, child => {
+      return category.children.some(child => {
         return !child.is_hidden
       })
     },
     onToggleCategory(category, update = true) {
       category.isOpen = !category.isOpen
-      _.forEach(category.children, child => {
+      category.children.forEach(child => {
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:32]: 👍

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 21]
```diff
@@ -115,34 +93,88 @@ export default {
...
+    },
+    hasDatesRangeFilter(cat, selectedFilters) {
+      if (cat.external_id !== this.publisherDateExternalId) return false
+
+      const datesFilter = selectedFilters.find(
+        item => item.external_id === this.publisherDateExternalId
+      )
+
+      if (datesFilter && datesFilter.items) {
+        const items = datesFilter.items.filter(item => item !== null)
```
Comment (Ruben-Hartog) [inline portal/src/components/FilterCategories/FilterCategories.component.js:131]: some

File: portal/src/pages/search.vue [hunk 22]
```diff
@@ -57,7 +57,12 @@
       >
         <div class="search__filter">
           <div class="search__filter_sticky">
-            <FilterCategories v-model="search" />
+            <FilterCategories
+              v-model="search"
+              :filter-categories="getFilterCategories() || []"
```
[the remark this answers is not in the document]
Comment (kiruys) [reply portal/src/pages/search.vue:62]: toch wel. het lukt me niet om het anders op te lossen

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 16 above)
Comment (kiruys) [reply portal/src/components/FilterCategories/FilterCategories.component.js:121]: ✅

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 17 above)
Comment (kiruys) [reply portal/src/components/FilterCategories/FilterCategories.component.js:103]: ✅

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 18 above)
Comment (kiruys) [reply portal/src/components/FilterCategories/FilterCategories.component.js:65]: kun je me uitleggen waarom? volgens mij wordt het er alleen maar ingewikkelder op

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 19 above)
Comment (kiruys) [reply portal/src/components/FilterCategories/FilterCategories.component.js:46]: zie boven

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 18 above)
Comment (Ruben-Hartog) [reply portal/src/components/FilterCategories/FilterCategories.component.js:65]: Ja, het kan er natuurlijk ook niet in zetten (dat had ik even over het hoofd gezeten). Dan zou ik er eerder een object met key/values van maken. Dan kun je de key zetten met de nieuwe values (en als het al bestaat wordt het overschreven). Maar dat is weer een hele refactor van bestaande code, dus laat maar zitten.

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 21 above)
Comment (kiruys) [reply portal/src/components/FilterCategories/FilterCategories.component.js:131]: ✅

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 19 above)
Comment (Ruben-Hartog) [reply portal/src/components/FilterCategories/FilterCategories.component.js:46]: Yes, same. Hier zou die andere structuur ook voordeel hebben, maar niet nu 😉

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 23]
```diff
@@ -115,34 +93,88 @@ export default {
...
+      // Return categories that build the filter tree
+      const filteredCategories = this.filterCategories.filter(
+        cat => cat.is_hidden === false
+      )
+
+      const selectedItems = this.selectedFilters
+        .flatMap(filter => filter.items)
+        .filter(filter => filter !== null)
+
+      filteredCategories.map(cat => {
```
[the remark this answers is not in the document]
Comment (kiruys) [reply portal/src/components/FilterCategories/FilterCategories.component.js]: ik heb mijn best gedaan om het duidelijker te maken

File: portal/src/pages/search.vue [hunk 24]
```diff
@@ -185,6 +183,9 @@ export default {
       this.search.page = 1
       this.$store.dispatch('searchMaterials', Object.assign({}, this.search))
       this.$router.push(this.generateSearchMaterialsQuery(this.search))
+    },
+    getFilterCategories() {
+      return this.materials ? this.materials.filter_categories : []
```
Comment (fako) [inline portal/src/pages/search.vue:188]: (this.materials && this.materials.filter_categories) ? this.materials.filter_categories : []
Dan hoeft de [] in de template niet meer lijkt mij op regel 62

**Comment by fako** [comment]
Ziet er goed uit! Paar kleine dingetjes nog. Vooral spread operator zou ik wat terughoudend mee zijn in loops. Spreads zijn in recente V8 een stuk sneller geworden, maar heb een donkerbruin vermoeden dat iOS niet volgt en in loops verwacht ik memory issues als gb niet snel genoeg is. Bovendien vind ik het soms leesbaarder zonder spread dan met. Zie comments.

File: portal/src/store/modules/materials.js [hunk 25]
```diff
@@ -10,6 +10,20 @@ import axios from '~/axios'
 
 const $log = injector.get('$log')
 
+function generateSearchParams(search) {
```
[the remark this answers is not in the document]
Comment (Ruben-Hartog) [reply portal/src/store/modules/materials.js:13]: Ja, klopt

File: portal/src/store/modules/materials.js [hunk 26]
```diff
@@ -10,6 +10,20 @@ import axios from '~/axios'
...
+  const filters = Object.keys(search.filters).reduce((memo, key) => {
```
[the remark this answers is not in the document]
Comment (Ruben-Hartog) [reply portal/src/store/modules/materials.js:14]: Scherp, volgens mij bouwde ik eerst een Object en kwam erachter dat het hier niet moest. Vergeten er een map van te maken.

File: service/surf/apps/filters/migrations/0014_auto_20200702_1005.py [hunk 27]
```diff
@@ -0,0 +1,20 @@
+# Generated by Django 2.2.13 on 2020-07-02 10:05
+
+from django.db import migrations
+
+
+def remove_double_yes_filter_item(apps, schema_editor):
+    MpttFilterItem = apps.get_model('filters', 'MpttFilterItem')
+    yes_filter = MpttFilterItem.objects.filter(name='yes')[:1].values_list("id", flat=True)
```
[the remark this answers is not in the document]
Comment (Ruben-Hartog) [reply service/surf/apps/filters/migrations/0014_auto_20200702_1005.py:8]: Reden dat ik het zo deed was omdat je nooit zeker weet of er niet toevallig maar 1 is. Dus ik verwijder alles, behalve 1. En het moet ook op acc. Dus dit leek me de meest zekere manier dat het overal gebeurt 😉 .

File: portal/src/store/modules/filter-categories.js [hunk 28]
```diff
@@ -120,22 +120,17 @@ export default {
...
-          items: _.reject(_.map(items, 'external_id'), _.isEmpty)
-        }
-      })
+
+      let filterMap = {}
+      for (const group in selectedGroups) {
+        const items = selectedGroups[group].map(item => {
+          return item.external_id
+        })
+        filterMap = { ...filterMap, [group]: items }
```
[the remark this answers is not in the document]
Comment (Ruben-Hartog) [reply portal/src/store/modules/filter-categories.js:131]: Dat zou ik ook zeggen, dan kan die `filterMap` ook een `const` zijn.

File: portal/src/pages/search.vue (hunk 24 above)
Comment (Ruben-Hartog) [reply portal/src/pages/search.vue:188]: Ja, dat dacht ik ook. Blijkt toch niet zo te zijn, omdat het een global store is. Dus `this.materials` kan al gevuld zijn uit een andere pagina (zonder filter_categories) waardoor er toch dingen fout gaan. Hier zou dan een extra check bij moeten of die filter_categories bestaat.

File: portal/src/components/Search/Search.component.js [hunk 29]
```diff
@@ -107,6 +106,13 @@ export default {
           return item
         }
       )
+    },
+    collectFilters() {
+      let filterMap = {}
+      this.$store.getters.search_filters.map(item => {
```
[the remark this answers is not in the document]
Comment (Ruben-Hartog) [reply portal/src/components/Search/Search.component.js:112]: Ik zou er een reduce van maken 🤓

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 15 above)
Comment (Ruben-Hartog) [reply portal/src/components/FilterCategories/FilterCategories.component.js:97]: Ja, denk het wel. Ik wil het even meepakken in die 'bol.com' filters als het niet teveel werk is.

File: portal/src/components/FilterCategories/FilterCategories.component.js [hunk 30]
```diff
@@ -1,111 +1,83 @@
...
+    childExternalIds(categoryId, itemId) {
+      const category = this.filterCategories.find(
+        category => category.external_id === categoryId
+      )
+      const item = category.children.find(item => item.external_id === itemId)
+      const iterator = (memo, item) => {
+        if (item.children.length > 0) {
+          item.children.forEach(child => iterator(memo, child))
         }
+        return [...memo, item.external_id]
```
[the remark this answers is not in the document]
Comment (Ruben-Hartog) [reply portal/src/components/FilterCategories/FilterCategories.component.js:40]: Het feit dat je memory problemen gaat krijgen hiervan ben ik het niet mee eens 😄 . Memo is alleen gedefinieerd binnen deze functie dus de garbage collector kan dit heel snel opruimen. Daar ga je echt geen memory problemen van krijgen. Eerder performance problemen, want de spread operator is trager dan bijvoorbeeld een push.

Het is een geneste children structuur, vandaar dat het een reduce is. Kan ook met forEach, maar recursie met een forEach maakt het in mijn ogen niet duidelijker.

File: portal/src/pages/search.vue (hunk 24 above)
Comment (fako) [reply portal/src/pages/search.vue:188]: Pfff, ok, gotcha

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 15 above)
Comment (fako) [reply portal/src/components/FilterCategories/FilterCategories.component.js:98]: 👍

File: portal/src/components/FilterCategories/FilterCategories.component.js (hunk 30 above)
[the remark this answers is not in the document]
Comment (fako) [reply portal/src/components/FilterCategories/FilterCategories.component.js]: GC kicked pas in na de functie, dus met enkele honderden items gaat het zeker optellen. Rond 5000 elements in mem was wel zo'n beetje max van oudere iPhones (toen ik duizenden elementen in een app moest verwerken voor hybrid app). In dit geval kom je daar never ever aan, want het is een query parse. Dus ja, kan wel.

[the remark this answers is not in the document]
Comment (Ruben-Hartog) [reply portal/src/components/FilterCategories/FilterCategories.component.js]: Ja, je hebt wel gelijk, want het is natuurlijk recursief 🙊 . Ik heb er trouwens een push van gemaakt, dus problem solved 😛

**Pull request merged**